ocaml/xapi/xapi_vm_snapshot.ml

```diff
@@ -223,6 +226,11 @@ let checkpoint ~__context ~vm ~new_name =
   then begin
     let localhost = Helpers.get_localhost ~__context in
     Db.VM.set_resident_on ~__context ~self:vm ~value:localhost;
+    match vgpu_records with
+    | [] -> ()
+    | vgpu_rec :: _ ->
+      let vgpu = Db.VGPU.get_by_uuid ~__context ~uuid:vgpu_rec.Db_actions.vGPU_uuid in
+      Db.VGPU.set_resident_on ~__context ~self:vgpu ~value:vgpu_rec.Db_actions.vGPU_resident_on;
```

Comment:
There is a race window now, where the pGPU can be stolen from the VM that is being checkpointed by another VM starting up. This is because after the `suspend` call returns (L209), and during the `clone`, the `VPGU.resident_on` field is clear.

In fact, the same thing is already true for memory, due to `VM.resident_on` being cleared at the same time.

How about addressing both of these in `force_state_reset_keep_current_operations`, by having that function recognise that a checkpoint is in progress, and if so, not clear those fields? I think we can determine this from the `current_operations` and then update the if-statement in https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_vm_lifecycle.ml#L640.

Reply:
Thanks, Rob. @robhoes  That makes sense.
I had updated the PR base on your comment, please help to review again. Thanks so much.